File: android/app/src/main/java/com/calcplus/calculator/core/disguise/DisguiseSurfaceHost.kt
package com.calcplus.calculator.core.disguise

import androidx.compose.runtime.Composable
import androidx.compose.runtime.LaunchedEffect
import androidx.compose.runtime.remember
import androidx.compose.ui.Modifier
import com.calcplus.calculator.core.lock.TokenRecorder

/**
 * The one host object per surface instance (§1.7): it owns the
 * [TokenRecorder], translates the face's event stream into a
 * `commit(tokens, overflowed)` for the state machine, and implements the
 * §1.1 overt-face buffer clear.
 *
 * The face itself never sees a buffer, a length rule or a verification result.
 * Nothing in here is logged.
 *
 * @param onInput called for every non-commit event, so the caller can apply the
 *   §1.3 revert rule (a `WRONG_CODE` caption reverts to `PROMPT_CURRENT` on the
 *   next token / clear / removeLast, never on a commit).
 */
@Composable
fun DisguiseSurfaceHost(
    face: DisguiseProvider,
    mode: DisguiseMode,
    caption: CaptionState?,
    failedAttemptToken: Int,
    onCommit: (tokens: List<String>, overflowed: Boolean) -> Unit,
    modifier: Modifier = Modifier,
    onInput: () -> Unit = {},
) {
    val recorder = remember { TokenRecorder() }

    // §1.1: whenever the pulse bumps for an OVERT face, the buffer and the
    // overflow flag go with it. The face visibly resets its entry, so anything
    // typed during the verification window must not linger in a buffer the face
    // no longer depicts. The calculator (covert) keeps its buffer in
    // verifyCurrent — that is iteration-1 behavior and it stays.
    //
    // Keyed on the token, so it skips nothing on the initial composition beyond
    // one harmless clear of an already-empty buffer.
    LaunchedEffect(failedAttemptToken) {
        if (!face.isCovert) recorder.clear()
    }

    face.Surface(
        mode = mode,
        caption = caption,
        failedAttemptToken = failedAttemptToken,
        events = { event ->
            val commit = recorder.handle(event)
            if (commit != null) {
                onCommit(commit.tokens, commit.overflowed)
            } else {
                onInput()
            }
        },
        modifier = modifier,
    )
}

File: ios/SafeBox/Disguise/DisguiseSurfaceHost.swift
import SwiftUI

/// Holds the recorder outside SwiftUI's value-diffing so recording a token does
/// not invalidate the surface mid-drag (the pattern face emits tokens inside a
/// live gesture). `@State` still owns the instance, so it is created with the
/// surface and dies with it — the §1.5 re-instantiation guarantee is unchanged.
@MainActor
private final class TokenRecorderBox {
    var recorder = TokenRecorder()
}

/// The one object per surface instance that owns the `TokenRecorder`,
/// translates the face's event stream, and calls the host's
/// `commit(tokens:overflowed:)` (decisions §1.7).
///
/// The §1.1 **overt buffer clear** lives here: when the host bumps the pulse
/// for an overt face, the face visibly resets its entry, so the buffer it no
/// longer depicts is cleared in the same step. The calculator is covert, so its
/// `verifyCurrent` buffer is deliberately left alone — that is iteration 1's
/// behavior and it stays.
struct DisguiseSurfaceHost: View {
    let disguise: any DisguiseProviding
    let mode: DisguiseMode
    let caption: LockBanner?
    let failedAttemptCount: Int
    /// Raw stream, for the host's caption-revert rule. Never inspect tokens.
    var onEvent: (DisguiseEvent) -> Void = { _ in }
    let onCommit: (_ tokens: [String], _ overflowed: Bool) -> Void

    @State private var box = TokenRecorderBox()

    var body: some View {
        disguise.makeSurface(
            mode: mode,
            caption: caption,
            failedAttemptCount: failedAttemptCount
        ) { event in
            onEvent(event)
            if let commit = box.recorder.apply(event) {
                onCommit(commit.tokens, commit.overflowed)
            }
        }
        .onChange(of: failedAttemptCount) { old, new in
            guard new > old, !disguise.isCovert else { return }
            box.recorder.clear()
        }
    }
}
